### ISSUE TITLE
detach() and clear()

### ISSUE THREAD

**Issue by gavinking** [hibernate/hibernate-rx#23]
`RxSession` needs `detach()` and `clear()` methods, which are occasionally useful for controlling memory usage.

Now, AFAIR, `detach()` doesn't imply any sort of cache invalidation, it's a purely in-memory operation, and so do we even need to do anything special at all with these operations? Can't they just delegate straight through to the blocking implementation? @Sanne @FroMage?

**Comment by FroMage** [comment]
Never used those methods, so this can definitely be done last.

**Comment by gavinking** [comment]
> Never used those methods, so this can definitely be done last.

Sure, but my point is I think it's totally trivial. Unless @Sanne corrects me.

**Comment by Sanne** [comment]
Having clear is clearly important, but please see the StateControl
interface.

Regarding detach I'm not super interested, could we defer adding such
additional methods?

I'd like to have a conversation with you all about detach/merge .. should
we really have them?

**Comment by gavinking** [comment]
> Having clear is clearly important, but please see the `StateControl` interface.

Well I have seen it but:

1. I don't see why we need to separate interface for these operations. Neither Hibernate nor JPA segregates them onto a different interface, and I can't guess what the motivation for this might be. 
2. `StateControl` is currently unimplemented, leaving me to guess at the intended implementation. Since `clear()` is declared `void`, I imagine that it is the implementation I have in mind which just clears the persistence context in a blocking way (since `clear()` doesn't affect the second-level cache).

> Regarding detach I'm not super interested, could we defer adding such additional methods?

Well, if I'm right, and `detach()` doesn't affect the second-level cache, then, like `clear()` it's just a `void` (blocking) method, and I can add it in < 5 minutes of work.

> I'd like to have a conversation with you all about detach/merge .. should we really have them?

Why do you think we shouldn't?

**Comment by Sanne** [comment]
I'm fine to remove `StateControl`. The idea was to experiment with separating the more advanced API for managing the state of managed entities / Session state from the more common API, but it's just an  idea.

Regarding detach / merge : in my experience people either:
- don't use these
- don't understand these, and get in trouble

TBH even myself I never use these, working with a mix of attached and detached objects is too confusing - but I appreciate that's my personal experience, we can talk about it.

In short, I was thinking it would be best to leave out any non-essential method: even if it's just 5 minutes to code for you, it's going to consume more than 5 minutes to support them in the future :)

**Comment by Sanne** [comment]
to clarify:

People can still work with detached objects even without such APIs: they can still close the source Session (or clear it) - either way it's slightly simpler as the graph won't contain a mix of managed and unmanaged elements.

For re-attaching, I have a similar sentiment: the semantics regarding update/refresh are subtle, while it's straight forward for a user to load a fresh copy and explicitly set what they wat to be set.

**Comment by gavinking** [comment]
@Sanne I think it's more work to explain to people why the API arbitrarily doesn't support things that the regular Hibernate `Session` and `EntityManager` APIs *do* support.

**Comment by gavinking** [comment]
@Sanne P.S. I understand why you don't like these operations, because I resisted adding them for *years*. But let me promise you that if you don't have them, people will complain and whine.

**Comment by Sanne** [comment]
ok you have a point there. Let's remove Merge from ORM core too :-P

**Comment by Sanne** [comment]
> @Sanne P.S. I understand why you don't like these operations, because I resisted adding them for _years_. But let me promise you that if you don't have them, people will complain and whine.

ah glad to hear.

ok so what about you don't add them yet either?

We know they are trival to expose later (as it seems you checked already?), but let's wait for some persuasive demand for them?

@emmanuelbernard and @sebersole ?

**Comment by gavinking** [comment]
@Sanne The purpose of this project is to make Hibernate work in a reactive programming environment. It is absolutely a non-goal to subset the Hibernate APIs, which almost all exist for good reasons. 

If we want to argue about how to make Hibernate usage more idiot proof, then that is the purpose of the Panache project, *not* of *this* project.

**Comment by gavinking** [comment]
There, they're added.

**Comment by Sanne** [comment]
hum, ok

Not meaning to argue about Hibernate APIs but I had the impression we needed to scope feature aggressively.

Thanks for adding them - just keep in mind we won't add all things we inherit from ORM core, at least not at the first iteration.

**Comment by gavinking** [comment]
> I had the impression we needed to scope feature aggressively.

Well we do need to do that for features which are hard to implement. But any operations which don't require interaction with the database or cache can be just trivially propagated straight through to the `RxSession` API.

Anyway, I've done most of them now, the only remaining ones are the stuff for filter enablement #26, but I think that's trivial too.

**Comment by gavinking** [comment]
I'm going to call this done.
